### ISSUE TITLE
Hot water switch is not working to turn on hot water

### ISSUE THREAD

**Issue by omarassil** [lukasroegner/homebridge-tado#16]
The hot water switch only works to switch off hot water but the ON state is not working. when turning on, it toggles off automatically after few seconds and doesn't change anything in Tado app

**Comment by lukasroegner** [comment]
Maybe @TeresaRoberts could help with this issue?

I went through the code and don't really understand this line:

https://github.com/lukasroegner/homebridge-tado-smart-thermostat/blob/be427676a071bdc7f6db54784497110a9c0f2e5d/src/tado-hotwater-zone.js#L94

The function signature in the `node-tado-client` looks like this:

```js
setZoneOverlay(home_id, zone_id, power, temperature, termination)
```

I think the temperature should be `null` instead of `hotWaterService.getCharacteristic(Characteristic.On).value`

**Comment by Cardo1** [comment]
This is rather confusing, as turning off the hot water only clears any overlay on the hot water zone. You can’t actually turn off the hot water, unless you originally turned it on via HK (where it sets an overlay).

Turning on sets an overlay of the pre-determined length.

**Comment by lukasroegner** [comment]
Yes, I also noticed that when reading through the code. I think the main issue is that a switch doesn't have an auto state, it can just be manual on or manual off. This doesn't map to the three states (auto, manual on, manual off) of the Tado API.

This means, that the current implementation can just do "manual on" and "auto", there is no "manual off".

**Comment by Cardo1** [comment]
> Yes, I also noticed that when reading through the code. I think the main issue is that a switch doesn't have an auto state, it can just be manual on or manual off. This doesn't map to the three states (auto, manual on, manual off) of the Tado API.
> 
> This means, that the current implementation can just do "manual on" and "auto", there is no "manual off".

Exactly.  This appeared to be the best compromise.

**Comment by lukasroegner** [comment]
> Exactly. This appeared to be the best compromise.

I agree, it is the best solution for what HomeKit offers. I already played around with the HeaterCoolerService and ThermostatService, all of them require a target temperature.

However, that doesn't explain why it is not working for @omarassil. It could be that the Tado API has issues with handling the `true` value for temperature (the issue I explained above).

**Comment by omarassil** [comment]
Hi @lukasroegner and @Cardo1 
Let me re-state the problem after I did more tests:
- When the hot water is ON in Tado, I click the button to turn off in Homekit and it immediately goes back ON
- when the hot water is OFF in Tado, I click the button to turn on in Homekit and it goes back OFF after maximum of 5 seconds (the polling interval) without changing anything in Tado
I attach a video to show a demo of this. Hope this helps: [https://drive.google.com/file/d/1gBSUAFtAUT3NPAF_LBUYzehIqZQ4cSup/view?usp=sharing](https://drive.google.com/file/d/1gBSUAFtAUT3NPAF_LBUYzehIqZQ4cSup/view?usp=sharing)

**Comment by lukasroegner** [comment]
@omarassil There is a target temperature in the Tado UI? How does that work?

**Comment by omarassil** [comment]
@lukasroegner yes this how it shows up in my tado app. As long as I set it to any value it will turn the hot water on. I believe it could support setting a specific temperature but it needs support from the hot water tank thermostat to send this info back to the control unit but for me I just use any temperature value and it works.

**Comment by Cardo1** [comment]
Watching the video, I’m not sure the plugin will work with what you’re trying.

In the first example, you’re trying to turn off the water which is on as per the schedule. The plugin can’t do this due to HK switch limitations. Turning water “off” in HK only clears an overlay. 

In the second part of the video, you turn off the water manually in tado, which is setting an overlay. You are then trying to set another overlay through HK to turn it on again.  The tado api may not accept this. Further testing is required. 

Can you have the water off in tado using the schedule, then try to turn it on using HK and see what happens?

**Comment by omarassil** [comment]
@Cardo1 thanks but even when it water off in Tado using the schedule and I turn on in HK it toggles off within 5 second (the polling duration). see new video below showing this case
https://drive.google.com/file/d/1vbUFqzwwSzOn3Nc5rcq9n5sGJzMfnnBl/view?usp=sharing

**Comment by Cardo1** [comment]
> @Cardo1 thanks but even when it water off in Tado using the schedule and I turn on in HK it toggles off within 5 second (the polling duration). see new video below showing this case
> https://drive.google.com/file/d/1vbUFqzwwSzOn3Nc5rcq9n5sGJzMfnnBl/view?usp=sharing

Ok, thanks for confirming. 

My boiler doesn’t support water temperature as this is controlled by the hot water tank. So in tado it’s simply on or off. 

I think in your case tado may require a temperature value to accept the overlay. Let me see...

**Comment by Cardo1** [comment]
I’m afraid my lack of coding skills is letting me down.  The wife isn’t available to have a look at the moment, though @lukasroegner will probably know.  Try manually editing the tado-hotwater-zone.js file to hardcode a specific water temperature when setting an overlay and then try it.  It will be on line 94 as linked above, though I don’t know which bit needs replacing.  Maybe the .value?  Though that’s probably wrong.  If that is the issue, an optional hot water temperature could be added for those that require it.

**Comment by omarassil** [comment]
I tried to debug and found that the value of `hotWaterService.getCharacteristic(Characteristic.On).value` which is passed to  setZoneOverlay in the line above is returning "false" and causing the function to fail. I tried to change to pass true, an integer, or null but none of these helped and it's still failing. 
the clearZoneOverlay is working fine as in if I have set the hot water to ON on Tado while the schedule is off, it will turn it back to auto from HK when I turn off the button

**Comment by omarassil** [comment]
Anyone has any thoughts on this issue? I'm happy to do tests on my setup but I tried all things that came to my mind and it didn't help

**Comment by lukasroegner** [comment]
@omarassil Can you provide some additional information on how you configured your hot water device:
- Link to the product page
- Screenshots of the configuration of the device (it seems that temperature control can be enabled/disabled?)
- Screenshots of the UI when setting the mode/temperature
- Minimum/Maximum temperature that can be set
- You already mentioned that a temperature slider is shown: how does the device know the current temperature? If it does not know the current temperature, when does the device switch on (at which target temperature that you set in the Tado app)?

A solution I can think of is to add a configuration option so that you can expose your hot water device the same way as thermostats are exposed to HomeKit.

**Comment by omarassil** [comment]
@lukasroegner I'm not sure if that will fix it because I'm passing this function 
setZoneOverlay(home_id, zone_id, power, temperature, termination)
and I'm passing a temperature in a form on integer but it's still not working

Minimum/Maximum temperature that can be set: min 30 and max 65. It turned on from any value of this range. It actually doesn't know the current temperature. The temperature at the hot water cylinder is controlled by a manual thermostat so any value passed here is ignored and it just turns on or off

Here is a video showing config and some screenshots below. 
Video: https://drive.google.com/file/d/15zJe15yVDbptBEB1qH0rSmMKGQ-8ewox/view?usp=sharing

![image](https://user-images.githubusercontent.com/18346514/90246067-b59ac300-de2b-11ea-88ae-869cec3f9dc8.png)

![image](https://user-images.githubusercontent.com/18346514/90246116-c77c6600-de2b-11ea-871c-9bcac4e8a84f.png)

**Comment by Cardo1** [comment]
That’s strange.  You have the same boiler that we have.  If the boiler doesn’t know what the water temperature is, only whether it wants to be on or off, why does your tado app provide a temperature setting?  This sounds like an issue with your tado configuration?
With the same boiler, our tado app simply shows an on/off switch for the hot water zone.

**Comment by omarassil** [comment]
@Cardo1 do you have a separate hot water cylinder? Yea I don't know why the app has temperature settings for hot water while it's redundant. when I fitted the system it didn't work straightaway and I had to phone Tado support and they did some config from their end

**Comment by Cardo1** [comment]
> @Cardo1 do you have a separate hot water cylinder? Yea I don't know why the app has temperature settings for hot water while it's redundant. when I fitted the system it didn't work straightaway and I had to phone Tado support and they did some config from their end

We have a hot water cylinder with a regular thermostat. I’m guessing you have the same setup we have. I’d get onto Tado and ask them to modify your hot water zone to remove the temperature setting.

**Comment by adavison007** [comment]
I have the same issue, the hot water button only mirrors the hot water from the Tado app, if I turn the hot water on or off in the HomeKit app, it does not change the Tado app.   

@SeydX  had the same issue when he did the Tado platform app.... you need to cater for and have the in settings the type of hot water you have.... so I have a cylinder and thus can't control the temperature of the hot water from Tado... the hot water is controlled by the Tado thermostat and not the extension kit.... hope this helps you to fix it???

See Json from Tado platform App

        "BU1412FT01051088094208-0": {
            "active": true,
            "overrideMode": "auto",
            "autoTemp": 50,
            "ID": 0,
            "zoneName": "Hot Water",
            "zoneType": "HOT_WATER",
            "deviceType": "BU",
            "serial": "BU1412FT01051088094208",
            "type": "boiler",
            "canSetTemperature": false

ALSO

Please can you change the hot water from a switch to a Faucet picture (tap) when it isn't a temperature control, but just on or off control.

PS yours works better than the Platform plugin other than this issue.

**Comment by adavison007** [comment]
@lukasroegner Any updates or ideas on how to fix the above? If you can get the Hot water working correctly, you'll have the best Homekit Tado integration for Homebridge :) thanks in a advance and really appreciate your support.

**Comment by lukasroegner** [comment]
> .... you need to cater for and have the in settings the type of hot water you have.... so I have a cylinder and thus can't control the temperature of the hot water from Tado... the hot water is controlled by the Tado thermostat and not the extension kit.... hope this helps you to fix it???

The main issue is that a switch (or a faucet) service doesn't have an "auto" setting, which means there is no way to expose multiple settings (current state: on/off, target state: on/off/auto) with a single switch (or faucet).

The solution to this problem is to use a different service type or multiple services rather than a single switch.
You can play around with the HomeKit Accessory Simulator to try out different accessories/services. What looks promising to me is the HeaterCooler service, which can be configured the following way:
- Current state: Inactive,Heating
- Target state: Off, Auto, Heat

The only thing with the HeaterCooler service is that it requires a CurrentTemperature characteristic. One way to solve this may be setting a virtual current temperature based on the state (e.g. 0 degree when inactive, 100 degree when heating).

**Comment by adavison007** [comment]
> > .... you need to cater for and have the in settings the type of hot water you have.... so I have a cylinder and thus can't control the temperature of the hot water from Tado... the hot water is controlled by the Tado thermostat and not the extension kit.... hope this helps you to fix it???
> 
> The main issue is that a switch (or a faucet) service doesn't have an "auto" setting, which means there is no way to expose multiple settings (current state: on/off, target state: on/off/auto) with a single switch (or faucet).
> 
> The solution to this problem is to use a different service type or multiple services rather than a single switch.
> You can play around with the HomeKit Accessory Simulator to try out different accessories/services. What looks promising to me is the HeaterCooler service, which can be configured the following way:
> 
> * Current state: Inactive,Heating
> * Target state: Off, Auto, Heat
> 
> The only thing with the HeaterCooler service is that it requires a CurrentTemperature characteristic. One way to solve this may be setting a virtual current temperature based on the state (e.g. 0 degree when inactive, 100 degree when heating).

Thanks @lukasroegner Sounds promising, if you check the config above from the tado-platform plugin, within th epson you set the following:-

 "canSetTemperature": false 

If you say false you get a faucet/switch only, if you set true you get a thermostat... would this be a possible fix?

**Comment by lukasroegner** [comment]
> If you say false you get a faucet/switch only, if you set true you get a thermostat... would this be a possible fix?

No, that won't fix the fact that switches/faucets don't have an auto mode.

In order to implement a clean and working solution that can be understood by users of the plugin, a service type/service combination must be chosen that supports at least three states: ON/OFF/AUTO.

**Possible solution (1): HeaterCooler**
Pros: supports all features that are required (three states, differentiates between current state and target state)
Cons: the CurrentTemperature is a required characteristic, which means this value has to be "emulated"

**Possible solution (2): 2 Switches/Faucets**
Pros: supports all features that are required (three states, one switch for on/off, on switch for AUTO on/off)
Cons: one combination of the switches is useless (when the first switch is off, the second switch for AUTO on/off is useless)

**Comment by adavison007** [comment]
> > If you say false you get a faucet/switch only, if you set true you get a thermostat... would this be a possible fix?
> 
> No, that won't fix the fact that switches/faucets don't have an auto mode.
> 
> In order to implement a clean and working solution that can be understood by users of the plugin, a service type/service combination must be chosen that supports at least three states: ON/OFF/AUTO.
> 
> **Possible solution (1): HeaterCooler**
> Pros: supports all features that are required (three states, differentiates between current state and target state)
> Cons: the CurrentTemperature is a required characteristic, which means this value has to be "emulated"
> 
> **Possible solution (2): 2 Switches/Faucets**
> Pros: supports all features that are required (three states, one switch for on/off, on switch for AUTO on/off)
> Cons: one combination of the switches is useless (when the first switch is off, the second switch for AUTO on/off is useless)

Many thanks for the ideas, personally I prefer option 1, then in the json settings allow people to select the emulated temp if there boiler/install does not allow temp control?

**Comment by mpjalexander** [comment]
I also have a tado system connected to a boiler that has it’s own thermostat so is presented as a simple on/off. I’m currently using the Tado Platform plugin, but considering switching to this one given it appears to be more actively supported. @lukasroegner, did you make any decision on the two options you suggested? Would it be feasible to implement both and allow one to be chosen via the config?

**Comment by lukasroegner** [comment]
Closed in favor of #31.
